File: framework/lib/vbox.dart
part of uxml;

/**
 * Implements a container of UIElement(s) that stacks items vertically.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */

class VBox extends UIElementContainer {
  static ElementDef vBoxElementDef;
  static PropertyDefinition spacingProperty;

  VBox() {
  }

  /**
   * Sets/returns spacing between elements.
   */
  num get spacing {
    return getProperty(spacingProperty);
  }

  set spacing(num val) {
    setProperty(spacingProperty, val);
  }

  /** Overrides UIElement.onMeasure to return size of element. */
  bool onMeasure(num availableWidth, num availableHeight) {
    num yPos = 0.0;
    num maxWidth = 0.0;
    UIElement child;
    num space = spacing;

    int numChildren = childCount;
    for (int childIndex = 0; childIndex < numChildren; childIndex++) {
      if (childIndex != 0 && child.visible) {
        yPos += space;
      }
      child = childElements[childIndex];
      child.measure(availableWidth, availableHeight);
      yPos += child.measuredHeight;
      if (child.measuredWidth > maxWidth) {
        maxWidth = child.measuredWidth;
      }
    }
    setMeasuredDimension(maxWidth, yPos);
    return false;
  }

  /** Returns the number of children that can fit in this container. */
  int estimateCapacity() {
    int numChildren = childCount;
    if (numChildren == 0) {
      return 0;
    }
    num yPos = 0;
    UIElement child;
    num space = spacing;
    num targetWidth = layoutWidth;
    num targetHeight = layoutHeight;
    int capacity = 0;
    int childIndex = 0;
    num childWidth = 0;
    num childHeight = 0;

    while (yPos < targetHeight) {
      if (childIndex != 0) {
        yPos += space;
      }

      if (childIndex < numChildren) {
        child = childElements[childIndex++];
        childWidth = max(child.measuredWidth, childWidth);
        childHeight = max(child.measuredHeight, childHeight);
      }

      if (yPos + childHeight < targetHeight) {
        capacity++;
      }

      yPos += childHeight;
    }
    return capacity;
  }


  /** Overrides UIElement.onLayout. */
  void onLayout(num targetX, num targetY, num targetWidth, num targetHeight) {
    num yPos = 0.0;
    UIElement child;
    num space = spacing;
    int numChildren = childCount;

    for (int childIndex = 0; childIndex < numChildren; childIndex++) {
      if (childIndex != 0 && child.visible) {
        yPos += space;
      }
      child = childElements[childIndex];
      child.layout(0.0, yPos, targetWidth, child.measuredHeight);
      yPos += child.measuredHeight;
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => vBoxElementDef;

  /** Registers component. */
  static void registerVBox() {
    spacingProperty = ElementRegistry.registerProperty("spacing",
        PropertyType.NUMBER, PropertyFlags.RESIZE | PropertyFlags.RELAYOUT ,
        null, 0.0);
    vBoxElementDef = ElementRegistry.register("VBox",
        UIElementContainer.uielementcontainerElementDef, [spacingProperty],
        null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/VBox.java
package com.hello.uxml.tools.framework;

import java.util.EnumSet;

/**
 * Implements vertical layout container.
 *
 * @author ferhat
 *
 */
public class VBox extends UIElementContainer {

  /** Spacing Property Definition */
  public static PropertyDefinition spacingPropDef = PropertySystem.register("Spacing", Double.class,
      VBox.class, new PropertyData(0.0, EnumSet.of(PropertyFlags.Resize, PropertyFlags.Relayout)));

  @Override
  protected void onMeasure(double availableWidth, double availableHeight) {
    double yPos = 0;
    double maxWidth = 0;
    double spacing = getSpacing();
    UIElement child;

    int childCount = getChildCount();
    for (int childIndex = 0; childIndex < childCount; ++childIndex) {
      if (childIndex != 0) {
        yPos += spacing;
      }
      child = getChild(childIndex);
      child.measure(availableWidth, availableHeight);
      yPos += child.getMeasuredHeight();
      if (child.getMeasuredWidth() > maxWidth) {
        maxWidth = child.getMeasuredWidth();
      }
    }
    setMeasuredDimension(maxWidth, yPos);
  }

  @Override
  protected void onLayout(Rectangle layoutRect) {
    double yPos = 0;
    double spacing = getSpacing();
    UIElement child;
    Rectangle childRect;
    int childCount = getChildCount();
    for (int childIndex = 0; childIndex < childCount; ++childIndex) {
      if (childIndex != 0) {
        yPos += spacing;
      }
      child = getChild(childIndex);
      childRect = new Rectangle(0, yPos, layoutRect.width,
          child.getMeasuredHeight());
      child.layout(childRect);
      yPos += childRect.height;
    }
  }

  /**
   * Sets/returns spacing between children.
   */
  public void setSpacing(double value) {
    setProperty(spacingPropDef, value);
  }

  public double getSpacing() {
    return ((Double) getProperty(spacingPropDef)).doubleValue();
  }
}
